### DISCUSSION TITLE
Where to put a custom partition layout?

### DISCUSSION THREAD

**Discussion by Schlumpf** [OE4T/meta-tegra#1467] [Q&A]
According to https://github.com/OE4T/meta-tegra/wiki/Redundant-Rootfs-A-B-Partition-Support, I a have a custom redundant flash layout XML file. In my distro.conf I set it with:
```
USE_REDUNDANT_FLASH_LAYOUT = "1"
PARTITION_LAYOUT_TEMPLATE_REDUNDANT = "flash_mycustom.xml"
```
The file is derived from flash_l4t_t194_spi_emmc_p3668_rootfs_ab.xml for a Xavier NX with some extensions.

But how do I get the file into the image?

**Comment by joekale** [comment]
I needed a new recipe to include the file in the rootfs and a bbappends to the tegra-bootfiles recipe to set that file  to the `PARTITION_FILE` and/or `PARTITION_FILE_EXTERNAL` variable and also add a dependency on my recipe that includes the xml file. 

This is the `tegra-bootfiles_%.bbappends` as an example.
```
DEPENDS += "rauc-custom-flash-layout"
RAUC_PARTITION_FILE = "${STAGING_DATADIR}/rauc-flash-layout/flash_rauc.xml"
PARTITION_FILE:custom-xavier-nx = "${RAUC_PARTITION_FILE}"
PARTITION_FILE_EXTERNAL:custom-orin-nano = "${RAUC_PARTITION_FILE}"
```

**Reply by madisongh** [reply]
For nanbield and later branches, this should get easier to do.   There's a `tegra-storage-layout` recipe there that you can directly add a bbappend to, and supply your modified XML files by adding to the SRC_URI in the bbappend.

**Reply by albertbu** [reply]
Could we please have an example? It doesn't work either way

**Reply by dwalkes** [reply]
There's an example adding a data partition at https://github.com/mendersoftware/meta-mender-community/blob/4c1c2c144c9155bcca5cbd0f9d310eebef56d6c9/meta-mender-tegra/recipes-bsp/tegra-binaries/tegra-storage-layout-base_%25.bbappend but it's more complicated than just the simple replacement of the XML file.

Regarding
> It doesn't work either way

Perhaps share a minimally modified fork of https://github.com/OE4T/tegra-demo-distro showing what you've attempted to change unsuccessfully and details about the issues you found would be a good approach.  One benefit is that it would also help build the public example you are requesting.

**Comment by arrow53** [comment]
i'm interested in an A / B rootfs. I'd like to try `tegra-storage-layout`. But, my use case is I always want a data partition.

Is there an example layout xml that I can start with to try out the above?

**Reply by albertbu** [reply]
Hi, did you solve the issue? I'm facing the same problem. Thanks!

**Reply by DenizEDRS** [reply]
were you able to resolve this?

**Reply by dwalkes** [reply]
There are working examples in each branch of https://github.com/Trellis-Logic/kas-demos/blob/main/layers/meta-trellis-swupdate-rootfs-overlay-oe4t/recipes-bsp/tegra-binaries/tegra-storage-layout_%25.bbappend for jetson-orin-nano-devkit-nvme - you can pick the one which applies to your target branch and look at the commit history for the files to use with different MACHINEs.

**Comment by albertbu** [comment]
Hi
I'll have a look, and be more than happy to share my work when I'll
succeed.
Alas, I do not have enough time to work on the demo distro, but I'll share
everything when I'll succeed
Thanks

On Mon, Sep 2, 2024 at 6:36 PM Dan Walkes ***@***.***> wrote:

> There's an example adding a data partition at
> https://github.com/mendersoftware/meta-mender-community/blob/4c1c2c144c9155bcca5cbd0f9d310eebef56d6c9/meta-mender-tegra/recipes-bsp/tegra-binaries/tegra-storage-layout-base_%25.bbappend
> but it's more complicated than just the simple replacement of the XML file.
>
> Regarding
>
> It doesn't work either way
>
> Perhaps share a minimally modified fork of
> https://github.com/OE4T/tegra-demo-distro showing what you've attempted
> to change unsuccessfully and details about the issues you found would be a
> good approach. One benefit is that it would also help build the public
> example you are requesting.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/OE4T/meta-tegra/discussions/1467#discussioncomment-10522556>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABRZY4H3B5HERVCG4EOQWFTZUSAV7AVCNFSM6AAAAABCWW7YJKVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTANJSGI2TKNQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Albert Burbea
Hana Zemer 9
Tel Aviv 62917, Israel
Tel/Fax + 972-3-6044704
Mobile: +972-52-3541842

**Reply by dwalkes** [reply]
> I do not have enough time to work on the demo distro

To be clear, I'm not suggesting you specifically add a functional demo to the distribution, just that you start by using it as a way to test your changes in isolation from whatever else is custom in your layer and demonstrate whatever isn't working in a way which makes it easier to provide feedback.

**Comment by albertbu** [comment]
hi,
well, the mender example just replaces / adds a filename to a partition,
but it's not changing the size of the partition itself
Albert

On Mon, Sep 2, 2024 at 6:36 PM Dan Walkes ***@***.***> wrote:

> There's an example adding a data partition at
> https://github.com/mendersoftware/meta-mender-community/blob/4c1c2c144c9155bcca5cbd0f9d310eebef56d6c9/meta-mender-tegra/recipes-bsp/tegra-binaries/tegra-storage-layout-base_%25.bbappend
> but it's more complicated than just the simple replacement of the XML file.
>
> Regarding
>
> It doesn't work either way
>
> Perhaps share a minimally modified fork of
> https://github.com/OE4T/tegra-demo-distro showing what you've attempted
> to change unsuccessfully and details about the issues you found would be a
> good approach. One benefit is that it would also help build the public
> example you are requesting.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/OE4T/meta-tegra/discussions/1467#discussioncomment-10522556>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABRZY4H3B5HERVCG4EOQWFTZUSAV7AVCNFSM6AAAAABCWW7YJKVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTANJSGI2TKNQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Albert Burbea
Hana Zemer 9
Tel Aviv 62917, Israel
Tel/Fax + 972-3-6044704
Mobile: +972-52-3541842

**Comment by Lexmark-chad** [comment]
Per what @madisongh indicated you can simply .bbappend to `tegra-storage-layout` and specify your own custom layout.  Something like this:

```
FILESEXTRAPATHS:prepend := "${THISDIR}/${BPN}:"

SRC_URI:append = " file://external-flash.xml"

PARTITION_FILE_EXTERNAL = "${WORKDIR}/external-flash.xml"
```

**Reply by djusten** [reply]
> ```
> PARTITION_FILE_EXTERNAL = "${WORKDIR}/external-flash.xml"
> ```

I used `PARTITION_FILE = "${WORKDIR}/external-flash.xml"` instead.

**Reply by dwalkes** [reply]
It will depend on the `MACHINE`, if you are working with an nvme variant you'll want to use PARTITION_FILE_EXTERNAL, if you have an emmc variant like the AGX Orin and aren't booting from nvme you'll use PARTITION_FILE.
